src/main/java/com/patchworkmc/event/initialization/RegisterEventRegistrars.java

```diff
@@ -1,56 +1,59 @@
-
-			method.visitMethodInsn(Opcodes.INVOKESPECIAL, shimName, "<init>", "()V", false);
-			method.visitMethodInsn(Opcodes.INVOKEINTERFACE, "net/minecraftforge/eventbus/api/EventRegistrarRegistry", "registerStatic", "(Ljava/lang/Class;Ljava/util/function/Consumer;)V", true);
-		}
-
-		for (Map.Entry<String, String> entry : instanceEventRegistrars) {
-			String shimName = entry.getKey();
-			String baseName = entry.getValue();
-
-			method.visitFieldInsn(Opcodes.GETSTATIC, "net/minecraftforge/eventbus/api/EventRegistrarRegistry", "INSTANCE", "Lnet/minecraftforge/eventbus/api/EventRegistrarRegistry;");
-
-			method.visitLdcInsn(Type.getObjectType(baseName));
-
-			method.visitTypeInsn(Opcodes.NEW, shimName);
-			method.visitInsn(Opcodes.DUP);
-
-			method.visitMethodInsn(Opcodes.INVOKESPECIAL, shimName, "<init>", "()V", false);
-			method.visitMethodInsn(Opcodes.INVOKEINTERFACE, "net/minecraftforge/eventbus/api/EventRegistrarRegistry", "registerInstance", "(Ljava/lang/Class;Ljava/util/function/BiConsumer;)V", true);
+		// We store the instance as a local variable for more readable decompiled code.
+		method.visitFieldInsn(Opcodes.GETSTATIC, EVENT_REGISTRAR_REGISTRY, "INSTANCE", EVENT_REGISTRAR_REGISTRY_INSTANCE_SIGNATURE);
+		method.visitVarInsn(Opcodes.ASTORE, 1);
+
+		for (EventSubscriber subscriber : eventSubscribers) {
+			String className = subscriber.getClassName();
+
+			if (subscriber.hasInstanceSubscriber()) {
+				method.visitVarInsn(Opcodes.ALOAD, 1); // Push the instance to the stack (1)
+				method.visitLdcInsn(Type.getObjectType(className)); // Push the class to the stack (2)
+				LambdaUtil.visitBiConsumerStaticMethodReference(method, className, EventConstants.REGISTER_INSTANCE, EventConstants.getRegisterInstanceDesc(className), subscriber.isInterface());
+				// Pop the instance for calling, and then pop the class lambda as parameters
+				method.visitMethodInsn(Opcodes.INVOKEINTERFACE, EVENT_REGISTRAR_REGISTRY, "registerInstance", "(Ljava/lang/Class;Ljava/util/function/BiConsumer;)V", true);
+			}
+
+			if (subscriber.hasStaticSubscriber()) {
+				method.visitVarInsn(Opcodes.ALOAD, 1); // Push the instance to the stack (1)
+				method.visitLdcInsn(Type.getObjectType(className)); // Push the class to the stack (2)
+				// Push the lambda to the stack (3)
+				LambdaUtil.visitConsumerStaticMethodReference(method, className, EventConstants.REGISTER_STATIC, EventConstants.REGISTER_STATIC_DESC, subscriber.isInterface());
+				// Pop the instance for calling, and then pop the class lambda as parameters
+				method.visitMethodInsn(Opcodes.INVOKEINTERFACE, EVENT_REGISTRAR_REGISTRY, "registerStatic", "(Ljava/lang/Class;Ljava/util/function/Consumer;)V", true);
+			}
 		}
 
+		Label end = new Label();
+		method.visitLabel(end);
 		method.visitInsn(Opcodes.RETURN);
-
-		method.visitMaxs(4, 0);
+		method.visitLocalVariable("registryInstance", EVENT_REGISTRAR_REGISTRY_INSTANCE_SIGNATURE, null, new Label(), new Label(), 1);
```

Comment:
Do the labels do anything here? I assume you'd want to pass the `start` and `end` labels as parameters?

Reply:
they're not needed, i tested in fernflower.
Will remove the dead label code.